### PR TITLE
fix(api): add stats to inbound lambda

### PR DESCRIPTION
**Pull request by Orta21**
Ref. metriport/metriport-internal#1669

Ticket: #1889 

### Dependencies

- Upstream: https://github.com/metriport/metriport-internal/pull/1809
- Downstream: none

### Description

Add analytics to inbound to determine how often we respond with docs after Patient Match

### Testing

- Staging
  - [ ] Infra works accordingly
  - [ ] Able to see analytics for inbound requests
- Production
  - [ ] Infra works accordingly
  - [ ] Able to see analytics for inbound requests

### Release Plan
- [ ] Upstream PR is merged
- [ ] Release NPM packages
- [ ] Merge this

### CODE REVIEW

File: packages/api/src/command/medical/document/document-conversion-status.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ export async function calculateDocumentConversionStatus({
     });
 
     if (isConversionCompleted) {
-      const startedAt = updatedPatient.data.documentQueryProgress?.startedAt;
+      const startedAt = externalData?.documentQueryProgress?.startedAt;
```
Comment (Orta21) [inline packages/api/src/command/medical/document/document-conversion-status.ts:73]: Forgot about this but its affected by scheduling the DQ so need to get the HIE startAt

File: packages/ihe-gateway-sdk/src/models/document-query/document-query-responses.ts [hunk 2]
```diff
@@ -17,7 +17,9 @@ export type InboundDocumentQueryRespSuccessful = z.infer<
   typeof inboundDocumentQueryRespSuccessfulSchema
 >;
 
-const inboundDocumentQueryRespFaultSchema = baseErrorResponseSchema;
+const inboundDocumentQueryRespFaultSchema = baseErrorResponseSchema.extend({
+  extrinsicObjectXmls: z.never().or(z.literal(undefined)),
```
Comment (Orta21) [inline packages/ihe-gateway-sdk/src/models/document-query/document-query-responses.ts:21]: If i dont set as never i cant extract from result

File: packages/ihe-gateway-sdk/src/models/document-retrieval/document-retrieval-responses.ts [hunk 3]
```diff
@@ -17,7 +17,9 @@ export type InboundDocumentRetrievalRespSuccessful = z.infer<
   typeof inboundDocumentRetrievalRespSuccessfulSchema
 >;
 
-const inboundDocumentRetrievalRespFaultSchema = baseErrorResponseSchema;
+const inboundDocumentRetrievalRespFaultSchema = baseErrorResponseSchema.extend({
+  documentReference: z.never().or(z.literal(undefined)),
+});
```
Comment (Orta21) [inline packages/ihe-gateway-sdk/src/models/document-retrieval/document-retrieval-responses.ts:22]: ditto

File: packages/infra/lib/ihe-stack.ts [hunk 4]
```diff
@@ -107,24 +113,32 @@ export class IHEStack extends Stack {
 
     const lambdaLayers = setupLambdasLayers(this, true);
 
+    const posthogSecretName = props.config.analyticsSecretNames?.POST_HOG_API_KEY;
+
     const documentQueryLambda = this.setupDocumentQueryLambda(
       props,
       lambdaLayers,
       vpc,
+      secrets,
```
Comment (RamilGaripov) [inline packages/infra/lib/ihe-stack.ts:122]: A couple comments here:

1. Instead of sending secrets + specific secret name for lookup, can't we just pass the retrieved secret value?
2. It's getting complicated with passing 6-7 arguments to these functions. It'll be very easy to mess the order of the parameters here. Can we wrap the parameters in an object?

**Review by RamilGaripov — APPROVED** [review]
Left a comment, which is kind of a nit... Address at your discretion 😄 

Also, would be good to release NPM packages after a PR is approved, so:
1. there are no iterations on NPM releases in case a change is requested in a published package, and
2. the pkg version updates don't litter the PR

File: packages/infra/lib/ihe-stack.ts [hunk 5]
```diff
@@ -170,12 +186,16 @@ export class IHEStack extends Stack {
...
         API_URL: props.config.loadBalancerDnsName,
+        ...(posthogSecretName ? { POST_HOG_API_KEY: posthogSecretName } : {}),
         ...(props.config.lambdasSentryDSN ? { SENTRY_DSN: props.config.lambdasSentryDSN } : {}),
       },
       vpc,
       alarmSnsAction,
       version: props.version,
     });
+
+    const posthogSecretKey = "POST_HOG_API_KEY";
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:197]: Why defining this multiple times instead of declaring a global const on this file?

File: packages/infra/lib/ihe-stack.ts [hunk 6]
```diff
@@ -195,12 +217,16 @@ export class IHEStack extends Stack {
       envType: props.config.environmentType,
       envVars: {
         MEDICAL_DOCUMENTS_BUCKET_NAME: props.config.medicalDocumentsBucketName,
+        ...(posthogSecretName ? { POST_HOG_API_KEY: posthogSecretName } : {}),
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:220]: This env var name is misleading, it doesn't contain the API key, but the secret name. Maybe `POSTHOG_API_KEY_SECRET_NAME`?

File: packages/infra/lib/ihe-stack.ts [hunk 7]
```diff
@@ -219,12 +247,17 @@ export class IHEStack extends Stack {
       envType: props.config.environmentType,
       envVars: {
         API_URL: props.config.loadBalancerDnsName,
+        ...(posthogSecretName ? { POST_HOG_API_KEY: posthogSecretName } : {}),
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:250]: ditto env var name

File: packages/infra/lib/ihe-stack.ts [hunk 8]
```diff
@@ -170,12 +186,16 @@ export class IHEStack extends Stack {
       envVars: {
         MEDICAL_DOCUMENTS_BUCKET_NAME: props.config.medicalDocumentsBucketName,
         API_URL: props.config.loadBalancerDnsName,
+        ...(posthogSecretName ? { POST_HOG_API_KEY: posthogSecretName } : {}),
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:189]: ditto env var name

File: packages/infra/lib/ihe-stack.ts [hunk 9]
```diff
@@ -195,12 +217,16 @@ export class IHEStack extends Stack {
...
         MEDICAL_DOCUMENTS_BUCKET_NAME: props.config.medicalDocumentsBucketName,
+        ...(posthogSecretName ? { POST_HOG_API_KEY: posthogSecretName } : {}),
         ...(props.config.lambdasSentryDSN ? { SENTRY_DSN: props.config.lambdasSentryDSN } : {}),
       },
       vpc,
       alarmSnsAction,
       version: props.version,
     });
+
+    const posthogSecretKey = "POST_HOG_API_KEY";
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:228]: ditto global const

File: packages/infra/lib/ihe-stack.ts [hunk 10]
```diff
@@ -219,12 +247,17 @@ export class IHEStack extends Stack {
...
         API_URL: props.config.loadBalancerDnsName,
+        ...(posthogSecretName ? { POST_HOG_API_KEY: posthogSecretName } : {}),
         ...(props.config.lambdasSentryDSN ? { SENTRY_DSN: props.config.lambdasSentryDSN } : {}),
       },
       vpc,
       alarmSnsAction,
       version: props.version,
     });
+
+    const posthogSecretKey = "POST_HOG_API_KEY";
```
Comment (leite08) [inline packages/infra/lib/ihe-stack.ts:258]: ditto global const

File: packages/api/src/shared/config.ts [hunk 11]
```diff
@@ -238,11 +238,6 @@ export class Config {
   static getCWMemberOID(): string {
     return getEnvVarOrFail("CW_MEMBER_OID");
   }
-
-  static getPostHogApiKey(): string | undefined {
-    return getEnvVar("POST_HOG_API_KEY");
-  }
-
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/shared/config.ts:245]: I see it was moved to core, but we need to add code to the lambda to get the value from the env var and call the AWS SDK to get the value of the secret w/ the env var content.

File: packages/infra/lib/ihe-stack.ts (hunk 4 above)
Comment (leite08) [reply packages/infra/lib/ihe-stack.ts:122]: > can't we just pass the retrieved secret value

No, secrets should not be stored on env vars.

> It's getting complicated with passing 6-7 arguments to these functions. It'll be very easy to mess the order of the parameters here. Can we wrap the parameters in an object?

😍

**Review by leite08 — CHANGES_REQUESTED** [review]

File: packages/lambdas/src/ihe-inbound-document-query.ts [hunk 12]
```diff
@@ -16,14 +19,23 @@ export const handler = Sentry.AWSLambda.wrapHandler(async (event: string) => {
...
-      distinctId: result.cxId,
-      event: EventTypes.inboundDocumentQuery,
-      properties: {
-        patientId: result.patientId,
-        documentCount: result.extrinsicObjectXmls.length,
+    let postHogApiKey: string | undefined;
+
+    if (postHogSecretName) {
+      postHogApiKey = await getSecretValue(postHogSecretName, region);
+    }
```
Comment (leite08) [inline packages/lambdas/src/ihe-inbound-document-query.ts:26]: Does it make sense to call analytics if there's no API key?

File: packages/core/src/util/config.ts [hunk 13]
```diff
@@ -91,4 +91,8 @@ export class Config {
   static getPostHogApiKey(): string | undefined {
     return getEnvVar("POST_HOG_API_KEY");
   }
+
+  static getUseMockS3Utils(): boolean {
+    return getEnvVar("USE_MOCK_S3_UTILS") === "true";
+  }
```
Comment (leite08) [inline packages/core/src/util/config.ts:97]: This is not part of this PR, is it?

File: packages/core/src/external/analytics/posthog.ts [hunk 14]
```diff
@@ -16,11 +16,12 @@ export interface EventMessageV1 extends IdentifyMessageV1 {
...
 
-const postApiKey = Config.getPostHogApiKey();
+const defaultPostHogApiKey = Config.getPostHogApiKey();
 
-export const analytics = (params: EventMessageV1) => {
-  if (postApiKey) {
-    const posthog = new PostHog(postApiKey);
+export const analytics = (params: EventMessageV1, postApiKey?: string) => {
+  const apiKey = postApiKey ?? defaultPostHogApiKey;
+  if (apiKey) {
```
Comment (leite08) [inline packages/core/src/external/analytics/posthog.ts:23]: nit: since we're touching this, could make this return if no api key available, instead of having the main flow inside an if

NABD, though

File: packages/core/src/util/config.ts (hunk 13 above)
Comment (Orta21) [reply packages/core/src/util/config.ts:97]: No ill remove

Comment (Orta21) [reply packages/core/src/util/config.ts]: Ill remove since its not being used anywere but seems like its already part of the code

Comment (leite08) [reply packages/core/src/util/config.ts]: Sounds good!

> seems like its already part of the code

FYI, that usually happens after a bad conflict resolution.

**Review by leite08 — APPROVED** [review]
